feat!: remove context from NoteEmission methods

nventuro opened the pull request:
In discussing how to improve developer experience by getting rid of the `context` object as much as possible, I realized some of it is quite low hanging fruit. We can take a similar approach to the `storage` object and have the aztec-nr intermediate structs keep references to the context, resulting in the end-user invoking methods that already reference the context instead of having to manually provide it. The diff speaks for itself.

Closes https://linear.app/aztec-labs/issue/F-109/remove-context-when-delivering-messages.

benesjan reviewed (approved):
Extremely nice change 🤩

Nico very smort

nventuro commented:
It turns out returning references (or structs holding references) from an if expression is not allowed. Now that `NoteEmission` contains such a reference, we run into the following:

```
error: Cannot return a reference type from an if or match expression
    ┌─ src/state_vars/private_mutable.nr:358:9
    │  
358 │ ╭         if !is_initialized {
359 │ │             self.initialize(init_note)
360 │ │         } else {
361 │ │             self.replace(transform_fn)
362 │ │         }
```

To address that, I created an internal auxiliary `NoteEmissionNoContext` type that can be created by calling `strip_context`, and have the context then re-attached via `attach_context`. The above snippet now reads

```noir
if !is_initialized {
    self.initialize(init_note).strip_context()
} else {
    self.replace(transform_fn).strip_context()
}
    .attach_context(self.context)
```

This is fine to do because the context in `self` is the same one as in the emissions in both branches (because it is a singleton).

nventuro commented:
I updated `OuterNoteEmission` to avoid the problematic mut ref on a return expression. The result is a bit nasty, but I want to get this out now - once we do a pass over `NoteEmission` and that API we'll also address `Outer`, its naming and how it's created, fixing this.

nventuro committed:
feat!: remove context from NoteEmission methods

In discussing how to improve developer experience by getting rid of the `context` object as much as possible, I realized some of it is quite low hanging fruit. We can take a similar approach to the `storage` object and have the aztec-nr intermediate structs keep references to the context, resulting in the end-user invoking methods that already reference the context instead of having to manually provide it. The diff speaks for itself.

Closes https://linear.app/aztec-labs/issue/F-109/remove-context-when-delivering-messages.

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>
